models/base/document.py

```diff
@@ -0,0 +1,231 @@
+
+        user = User(username='admin', password_hash='1321A7FE01')
+
+    To customize class initialization, override the __init__ method. For example:
+
+        def __init__(self, username, password):
+            self.validate_username(username)
+            password_hash = self.hash_password(password)
+            super().__init__(username=username, password_hash=password_hash)
+
+    After updating the document, call the save method to persist the change. For example:
+
+        user.username = 'admin2'
+        user.save()
+
+    To delete the document, use the delete method. For example:
+
+        user.delete()
+
+    Deleting a document also automatically removes all references to it.
+    """
+    _initialised = False
+
+    class PrimaryKeyNotDefinedError(Exception):
+        def __init__(self, document):
+            super().__init__(f"Primary key not defined for {document.__class__.__name__}")
+
+    class PrimaryKeyNotSetError(Exception):
+        def __init__(self, field_name):
+            super().__init__(f"Primary key {field_name} not set")
+
+    class ReferrerNotFound(Exception):
+        def __init__(self, referrer_type=None, attribute_name=None):
+            super().__init__(f"Instanced is not referenced by "
+                             f"{attribute_name or 'any field'} in {referrer_type or 'Any type'}")
+
+    @persist
+    def __init__(self, **kwargs):
+        """
+        Initialize the document.
+        All fields are assigned automatically.
+        :param kwargs: value to each field
+        """
+        self._data = {}
+        for field_name, field in self._fields.items():
+            if field.primary_key and field_name not in kwargs:
+                raise self.PrimaryKeyNotSetError(field_name)
+            self.__setattr__(field_name, kwargs.get(field_name))
+        self._initialised = True
+        self._referenced_by: list[Document] = []
+
+    def save(self) -> None:
+        """
+        Recursively call save on the root-level document
+        """
+        for referrer in self._referenced_by:
+            referrer.save()
+
+    def delete(self) -> None:
+        """
+        Remove all references to this document and persist the change.
+        """
+        for referrer in list(self._referenced_by):  # copy with list() to avoid issues with removing items
+            referrer.__unlink_referee(self)
+            referrer.save()
+
+    @property
+    def key(self):
+        """
+        The primary key value of the document.
+        """
+        return getattr(self, self._primary_key)
+
+    def find_referred_by(self, referrer_type: type = None, field_name: str = None) -> Document:
+        """
+        Find the first document that references this document matching the criteria.
+        ReferrerNotFound exception is raised if no referrer matching the criteria is found.
+        :param referrer_type: optional type of the referrer as a criteria
```

Comment:
Not clear on what "referrer_type" is.

Reply:
it is the class if the referrer for filtering referrers. 
e.g. 
```python
camp.find_referred_by(referrer_type=Plan)
```